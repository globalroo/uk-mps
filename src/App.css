.App-header {
  background-color: #222;
  padding: 20px;
  color: white;
  margin-bottom: 20px;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

.mp-image {
	object-fit: cover;
	object-position: top;
	max-height: 290px;
}

.image {
	background: url("404.jpeg") !important;
	height:290px;
}

.mp-container {
	display: flex;
	align-items: center;
	flex-direction: column;
}

.mp-animation-enclosure {
	display: flex;
	flex-direction: row;
	align-items: flex-start;
	width: 50%;
	overflow: hidden;
}

.extra-content {
	display: flex;
    align-items: center;
	justify-content: space-between;
	max-height:50px;
}

.mp-start-date {
	flex-grow:1;
}

.mp-animation-enclosure {
	display: flex;
	flex-direction: row;
	align-items: flex-start;
	width: 50%;
	overflow: hidden;
}

@media screen and (max-width: 414px) {
	.mp-animation-enclosure {
		width: 100%;
	}
}
