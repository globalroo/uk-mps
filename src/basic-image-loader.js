import React, { Component } from "react";
import PropTypes from "prop-types";

export class BasicImageLoader extends Component {
	constructor(props) {
		super(props);
		this.state = {
			error: false,
			loaded: false
		};
	}
	onLoad = () => {
		this.setState({
			loaded: true
		});
	}
	onError = () => {
		this.setState({
			error: true
		});
	}
	render() {
		const style = this.state.error ? { display: "none", opacity: 0 } : { opacity: (this.state.loaded ? 1 : 0 ) };
		return <img className="mp-image" style={style} src={this.props.src} onLoad={this.onLoad} onError={this.onError} alt={this.props.alt}/>;
	}
}

BasicImageLoader.propTypes = {
	src: PropTypes.string,
	alt: PropTypes.string,
	styles: PropTypes.object
};
