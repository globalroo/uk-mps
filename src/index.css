body {
  margin: 0;
  padding: 0;
  font-family: sans-serif;
  background-color: #ccc
}
.ui.card>.image, .ui.cards>.card>.image {
	flex:1;
}
