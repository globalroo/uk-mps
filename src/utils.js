const RDF_DESCRIPTOR = "http://www.w3.org/1999/02/22-rdf-syntax-ns#type";
const FULLNAME_BY_SURNAME_KEY = "http://example.com/A5EE13ABE03C4D3A8F1A274F57097B6C";
const FULLNAME_BY_FORENAME_KEY = "http://example.com/F31CBD81AD8343898B49DC65743F0BDF";
const MEMBER_IMAGE_KEY = "https://id.parliament.uk/schema/memberHasMemberImage";
const FAMILY_NAME_KEY = "https://id.parliament.uk/schema/personFamilyName";
const GIVEN_NAME_KEY = "https://id.parliament.uk/schema/personGivenName";
const PARLIAMENT_API = "http://lda.data.parliament.uk/members.json";
const LOCALSTORE_KEY = "membersOfParliament";
const PERSON_TYPE = "Person";

export const getThumbnailPortrait = (id = "404") => `images/${id}.jpeg`;
export const getTailOfKey = (key = "") => key.split("/").pop();
export const getCachedMembers = async () => await localStorage.getItem("membersOfParliament");

const getValueFromKeyPair = (keypairArr = []) => {
	const [keypair = { value: undefined }] = keypairArr;
	const { value } = keypair;
	return value;
};

export const getDataFromParliament = async name => {
	const response = await fetch(
		encodeURI(`${PARLIAMENT_API}?fullName=${name}`)
	);
	const jsonResponse = await response.json();
	const { result = { items: [] } } = jsonResponse;
	const { items = [] } = result;
	const [data] = items;
	return data;
};

export const flattenData = async members => {
	const formattedMembers = [];
	for (const [recordIndex, recordData] of Object.entries(members)) {
		const recordType = getValueFromKeyPair(recordData[RDF_DESCRIPTOR]);
		// Only find people - bunch of Nodes in the raw data
		if (recordType && recordType.includes(PERSON_TYPE)) {
			const mpKey = getTailOfKey(recordIndex);
			formattedMembers.push({
				id: getTailOfKey(mpKey),
				image: getThumbnailPortrait(
					getTailOfKey(getValueFromKeyPair(recordData[MEMBER_IMAGE_KEY]))
				),
				fullNameByForename: getValueFromKeyPair(recordData[FULLNAME_BY_FORENAME_KEY]),
				fullNameBySurname: getValueFromKeyPair(recordData[FULLNAME_BY_SURNAME_KEY]),
				familyName: getValueFromKeyPair(recordData[FAMILY_NAME_KEY]),
				givenName: getValueFromKeyPair(recordData[GIVEN_NAME_KEY]),
				data: await getDataFromParliament(getValueFromKeyPair(recordData[FULLNAME_BY_FORENAME_KEY]))
			});
		}
	}
	localStorage.setItem(LOCALSTORE_KEY, JSON.stringify(formattedMembers));
	return formattedMembers;
};
