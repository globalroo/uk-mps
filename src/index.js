import React from "react";
import ReactDOM from "react-dom";
import App from "src/App";
import registerServiceWorker from "src/registerServiceWorker";

import "semantic-ui-css/semantic.min.css";
import "src/index.css";
import members from "data/processed.json";

// Grab unique party names.
// const partyNames = members.map((rec) => rec.partyName);
// const unique = new Set([...partyNames]);

//TODO: Sort members alphabetically, move this to parser.
members.sort((a, b) => {
	if (a.personFamilyName < b.personFamilyName) return -1;
	if (a.personFamilyName > b.personFamilyName) return 1;
	return 0;
});

ReactDOM.render(<App members={members} />, document.getElementById("root"));
registerServiceWorker();
