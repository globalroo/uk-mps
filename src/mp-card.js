import React from "react";
import PropTypes from "prop-types";
import { format } from "date-fns";
import { BasicImageLoader } from "src/basic-image-loader";

const MISSING_IMAGE = "404";

const partyColours = {
	Conservative: "blue",
	Labour: "red",
	ScottishNationalParty: "yellow",
	Independent: "white",
	LiberalDemocrat: "orange",
	DemocraticUnionistParty: "orange",
	PlaidCymru: "lightGreen",
	GreenParty: "green",
	SinnFÃ©in: "darkOrange",
	IndependentUnionist: "lightBlue",
	UlsterUnionistParty: "blue",
	Speaker: "green"
};

export const getDate = (date) => format(date.split("+").shift(), "MMM YYYY");

export const getBestAvailableImage = (availableImage, twitterImage) => {
	let image = availableImage;
	if (image.includes(MISSING_IMAGE)) {
		if (twitterImage) {
			image = twitterImage;
		}
	}
	return image;
};
export const MPCard = ({ mp }) => {
	const index = mp.partyName.split(" ").join("");
	const color = partyColours[index];
	const classes = `${color} card card-overrides`;
	const availableImage = getBestAvailableImage(mp.localImage, mp.twitterImage);
	return (
		<div className={classes}>
			<div className="image image-container">
				<BasicImageLoader src={availableImage} alt={mp.honorificName} />
			</div>
			<div className="content">
				<div className="header">{mp.honorificName}</div>
				<div className="meta">{mp.constituencyGroupName}</div>
				<div className="description">{mp.positionName}</div>
			</div>
			<div className="extra content">
				<span className="right floated">
					<i className="calendar icon" />
					{getDate(mp.constituencyGroupStartDate)}
				</span>
				<span>
					<i className="id badge icon" />
					{mp.partyName}
				</span>
			</div>
		</div>
	);
};

MPCard.propTypes = {
	mp: PropTypes.object
};
