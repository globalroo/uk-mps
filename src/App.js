import React, { Component } from "react";
import PropTypes from "prop-types";
import "font-awesome/css/font-awesome.min.css";
import { MPCard } from "src/mp-card";

import "src/App.css";
import { CardStack } from "./card-stack/card-stack";import { CardStackControls } from "./card-stack/card-stack-controls";

class App extends Component {
	componentDidMount() {
		const options = {
			postAnimationCallback: this.enableControls
		};
		this.cardStack = new CardStack("stack_krisna",this.mpListRef, options);
		this.controls = new CardStackControls(this.cardStack);
	}

	enableControls = () => {
		this.controls.enable();
	}

	render() {
		return (
			<div className="App">
				<header className="App-header">
					<h1 className="App-title">Members of Parliament</h1>
					<p className="App-intro">
						Back to project:{" "}
						<a className="link" href="https://github.com/globalroo/uk-mps">
							uk-mps
						</a>{" "}
						- work in progress for memory game.
					</p>
				</header>
				<div className="mp-container">
					<div className="mp-animation-enclosure">
						<ul id="stack_krisna" ref={(ref) => this.mpListRef = ref} className="stack stack--krisna">

							{this.props.members.map((mp) => {
								return <li className="stack__item" key={`${mp.id}_mp_card_key_li`}>
									<MPCard mp={mp} key={`${mp.id}_mp_card_key`} />
								</li>;
							})}

						</ul>
					</div>
					<div className="controls">
						<button className="button button--sonar button--reject" data-stack="stack_krisna"><i className="fa fa-times"></i><span className="text-hidden">Reject</span></button>
						<button className="button button--sonar button--accept" data-stack="stack_krisna"><i className="fa fa-check"></i><span className="text-hidden">Accept</span></button>
					</div>
				</div>

			</div>
		);
	}
}

App.propTypes = {
	members: PropTypes.array
};

export default App;
