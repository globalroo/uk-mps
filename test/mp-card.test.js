import React from "react";
import * as renderer from "react-test-renderer";
import { MPCard, getBestAvailableImage } from "src/mp-card";
import members from "data/processed.json";

const singleMP = members.shift();

describe("MPCard tests", () => {
	test("MPCard is rendered as expected", () => {
		const component = renderer.create(<MPCard mp={singleMP} />);
		expect(component.toJSON()).toMatchSnapshot();
	});
	test("getBestAvailableImage returns a good image", () => {
		const testImage = "good.jpeg";
		const twitterImage = "twitter.jpeg";
		const availableImage = getBestAvailableImage(testImage, twitterImage);
		expect(availableImage).toBe("good.jpeg");
	});
	test("getBestAvailableImage returns a twitter image, if no default image", () => {
		const testImage = "404";
		const twitterImage = "twitter.jpeg";
		const availableImage = getBestAvailableImage(testImage, twitterImage);
		expect(availableImage).toBe("twitter.jpeg");
	});
	test("getBestAvailableImage leaves the 404 in place if no Twitter image", () => {
		const testImage = "404.jpeg";
		const twitterImage = "";
		const availableImage = getBestAvailableImage(testImage, twitterImage);
		expect(availableImage).toBe("404.jpeg");
	});
});
