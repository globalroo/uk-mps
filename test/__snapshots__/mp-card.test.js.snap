// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MPCard tests MPCard is rendered as expected 1`] = `
<div
  className="ui blue card card-overrides"
>
  <div
    className="image image-container"
  >
    <img
      alt="Dr Dan Poulter MP"
      className="mp-image"
      onError={[Function]}
      onLoad={[Function]}
      src="https://res.cloudinary.com/globalroo/image/twitter_name/w_300/drdanpoulter.jpeg"
      style={
        Object {
          "opacity": 0,
        }
      }
    />
  </div>
  <div
    className="content"
  >
    <div
      className="header"
    >
      Dr Dan Poulter MP
    </div>
    <div
      className="meta"
    >
      Central Suffolk and North Ipswich
    </div>
    <div
      className="description"
    >
      Parliamentary Under-Secretary (Department of Health)
    </div>
  </div>
  <div
    className="extra-content content"
  >
    <div
      className="extra-content-cell mp-start-date"
    >
      <i
        className="calendar icon"
      />
      May 2010
    </div>
    <div
      className="extra-content-cell mp-party-name"
    >
      <i
        className="id badge icon"
      />
      Conservative
    </div>
  </div>
</div>
`;
